resource-service/nats/connection.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/nats-io/nats.go"
+	logger "github.com/sirupsen/logrus"
+	"reflect"
+	"sort"
+)
+
+const streamName = "keptn"
+const consumerName = "resource-service"
+
+//go:generate moq --skip-ensure -pkg nats_mock -out ./mock/keptn_nats_handler_mock.go . IKeptnNatsMessageHandler
+type IKeptnNatsMessageHandler interface {
+	Process(event models.Event, sync bool) error
+}
+
+type processFunc func(event models.Event, sync bool) error
+
+type keptnNatsMessageHandler struct {
+	f processFunc
+}
+
+func NewKeptnNatsMessageHandler(f processFunc) *keptnNatsMessageHandler {
+	return &keptnNatsMessageHandler{
+		f: f,
+	}
+}
+
+func (nmh *keptnNatsMessageHandler) Process(event models.Event, sync bool) error {
+	return nmh.f(event, sync)
+}
+
+type NatsConnectionHandler struct {
+	natsConnection *nats.Conn
+	subscriptions  []*PullSubscription
+	topics         []string
+	natsURL        string
+	ctx            context.Context
+	jetStream      nats.JetStreamContext
+}
+
+func NewNatsConnectionHandler(ctx context.Context, natsURL string) *NatsConnectionHandler {
+	return &NatsConnectionHandler{natsURL: natsURL, ctx: ctx}
+}
+
+func (nch *NatsConnectionHandler) RemoveAllSubscriptions() {
+	for _, sub := range nch.subscriptions {
+		_ = sub.Unsubscribe()
+		logger.Infof("Unsubscribed from NATS topic: %s", sub.subscription.Subject)
+	}
+	nch.subscriptions = nch.subscriptions[:0]
+}
+
+// SubscribeToTopics expresses interest in the given subject(s) on the NATS message broker.
+func (nch *NatsConnectionHandler) SubscribeToTopics(topics []string, messageHandler IKeptnNatsMessageHandler) error {
+	if nch.natsURL == "" {
+		return errors.New("no PubSub URL defined")
+	}
+
+	if nch.natsConnection == nil || !nch.natsConnection.IsConnected() {
+		if err := nch.renewNatsConnection(); err != nil {
+			return err
+		}
+	}
+
+	if nch.jetStream == nil {
+		if err := nch.setupJetStreamContext(topics); err != nil {
+			return err
+		}
+	}
+
+	if len(topics) > 0 && !IsEqual(nch.topics, topics) {
+		nch.RemoveAllSubscriptions()
+		nch.topics = topics
+
+		for _, topic := range nch.topics {
+			subscription := NewPullSubscription(nch.ctx, "", topic, nch.jetStream, messageHandler.Process)
+			if err := subscription.Activate(); err != nil {
+				return fmt.Errorf("could not start subscription: %s", err.Error())
+			}
+			nch.subscriptions = append(nch.subscriptions, subscription)
+		}
+	}
+	return nil
+}
+
+func (nch *NatsConnectionHandler) GetPublisher() (*Publisher, error) {
+	if nch.natsConnection == nil || !nch.natsConnection.IsConnected() {
+		if err := nch.renewNatsConnection(); err != nil {
+			return nil, err
+		}
+	}
+	return NewPublisher(nch.natsConnection), nil
+}
+
+func (nch *NatsConnectionHandler) renewNatsConnection() error {
+	var err error
+	nch.RemoveAllSubscriptions()
+
+	nch.natsConnection.Close()
+	logger.Infof("Connecting to NATS server at %s ...", nch.natsURL)
+	nch.natsConnection, err = nats.Connect(nch.natsURL, nats.MaxReconnects(-1))
```

Comment:
can you please add the `nats.MaxReconnects(-1)` to the nats connection setup in the shipyard controller as well?

Reply:
done